[Wallet] Add support for 8 digits verification codes

i1skn opened the pull request:
### Description

This PR implements short verification codes. See the demo below.
To enable this feature we need all the validators to support `https`, which is not the case atm, so this feature are disabled by feature flag.

### Short code
Essentially, short code, which user receives consist of two parts: prefix and security code (e.g. `12345678`: `1` - prefix, `2345678` - security code).

We use prefix to identify which issuer has sent the code. Because we can not store persistent mapping of prefix<>issuer on the client side (will be gone after reinstall), we are doing the following:
* Set prefix to issuer's address mod 10 ([0,9]) and pass it with reveal request.
* When we receive a short code with a prefix back, we create a list of validators which map to the same prefix.
* We send security code to obtain attestation code to all issuers from the previous step
* One of them should return a full attestation code, all others should fail.

This approach allows us to scale for more than 10 actionable attestations at the same time and to avoid changing `Attestation.sol` contract to store mapping prefix<>issuer in it.

#### Caveat
We do occasionally redundant requests with security code to wrong validators and those exposing the code itself and increasing waiting time for the user. For 3 SMSs the probability of this happening is 30%.

For more information see this discussion https://discord.com/channels/600834479145353243/732977310101405736/778336853505081344
### Demo
![screen-20201119-212100~2](https://user-images.githubusercontent.com/6160124/99726972-426faf80-2af2-11eb-97b4-0909b8ded3ac.gif)

### Tested

With short codes enabled/disabled.

### Related issues

- Fixes #5384
- Fixes #5394

### Backwards compatibility

Yes

i1skn committed:
[Wallet] Add support for 8 digits verification codes

codyborn commented on packages/contractkit/src/wrappers/Attestations.ts:
```diff
@@ -67,6 +67,16 @@ export interface AttesationServiceRevealRequest {
+  securityCodePrefix?: string
+}
+
+export interface AttesationServiceCodeForSecurityRequest {
```
nit
```suggestion
export interface AttestationServiceCodeForSecurityRequest {
```

codyborn commented on packages/contractkit/src/wrappers/Attestations.ts:
```diff
@@ -583,6 +575,19 @@ export class AttestationsWrapper extends BaseWrapper<Attestations> {
+   */
+  getAttestationForSecurityCode(serviceURL: string, body: AttesationServiceCodeForSecurityRequest) {
+    const urlParams = new URLSearchParams({ ...body })
+    return fetch(appendPath(serviceURL, 'get_attestations') + '?' + urlParams, {
```
Does this include any retry logic? We should add this (with exponential backoff) if not.

i1skn replied:
I'm having a hard time to understand when do we want to do that. Basically the fact that we are not 100% sure that current validator is the one who sent the code brings some complexity here. Possible cases I see:

* Timeout - is it really worth trying it once again considering that it might not be the one we need.
* `4xx` - we expect this to happen for some validators.
* `5xx` - something went really wrong on the attestation service side. Probably, it make sense to try it later, but how much time we want to spend on each validator in total?

Any thoughts? cc: @aslawson @timmoreton @codyborn

codyborn replied:
Perhaps we can parallelize the requests.. if we get a 200 response from one of them we return immediately. Otherwise, for each 500 or timeout we retry a couple times before giving up.

aslawson replied:
This would be an optimization. We do have manual retry by the user.  If there is a failure and user selects "resend", it will try the existing actionable issuer first again. We allow the user to do this after 1 min, but if we receive either a timeout or a failure (other than expected failure on "already sent too many times" or "wrong code") we could resend the request within that minute.

aslawson replied:
actually i suppose that's for the initial request for get the security code, so really only "wrong code" would be the acceptable error

timmoreton replied:
I think it would be nice if we can retry with backoff except on cases where we get 4xx -- especially 403 (security code invalid, or attempts exceeded) or 404 (attestation not found -- i.e it was a hash collision and this is the wrong service).  However I feel it is an optimization if time is tight

codyborn commented on packages/mobile/src/identity/utils.test.ts:
```diff
@@ -0,0 +1,22 @@
+      '51365977'
+    )
+  })
+  it('should NOT extract with incorrect message', () => {
```
Sorry, I don't follow. Why is this an incorrect message when the one above is correct?

i1skn replied:
The security code consist of two parts: first digit is the prefix we use to identify the validator, when we receive the code and the rest 7 digits is the code itself. I will a comment there with the explanation as well 👍

codyborn commented on packages/mobile/src/identity/verification.ts:
```diff
@@ -618,13 +623,55 @@ export function attestationCodeReceiver(
+            (attestation) => hashAddressToSingleDigit(attestation.issuer) === lookupHash
+          )
+
+          for (const attestation of lookupAttestations) {
```
Perhaps we can wrap the code creation and verification in a library/class. For protocols like this, it helps to not expose these details to the caller since a future code change can break it if the author doesn't understand how the complete picture.

codyborn commented on packages/mobile/src/identity/verification.ts:
```diff
@@ -618,13 +623,55 @@ export function attestationCodeReceiver(
+                account,
+                phoneHashDetails,
+                attestation,
+                securityCode.substr(1)
```
Why is it removing the first char?

i1skn replied:
See the explanation above

codyborn commented on packages/mobile/src/identity/verification.ts:
```diff
@@ -618,13 +623,55 @@ export function attestationCodeReceiver(
-      if (!code) {
+      if (features.SHORT_VERIFICATION_CODES) {
+        securityCode = extractShortSecurityCodeMessage(action.message)
+        if (securityCode) {
```
Should we log in case securityCode is undefined?

codyborn commented on packages/mobile/src/identity/verification.ts:
```diff
@@ -618,13 +623,55 @@ export function attestationCodeReceiver(
+          }
+        }
+      } else {
+        attestationCode = extractAttestationCodeFromMessage(action.message)
```
It might be cleaner if we run this out of the else statement regardless and just change what we pass into it.
```
let message = action.message
if (features.SHORT_VERIFICATION_CODES) {
  ...
  message = yield call(...)
}
attestationCode = extractAttestationCodeFromMessage(message)
```

codyborn reviewed (changes requested):
Looks good! Added some suggestions about organization.

i1skn committed:
Address PR feedback

i1skn committed:
Fix env-test build

codyborn reviewed (approved):
Nice work! There's still the open question on the retry logic.. let me know what you think.

aslawson commented on packages/contractkit/src/wrappers/Attestations.ts:
```diff
@@ -67,6 +67,16 @@ export interface AttesationServiceRevealRequest {
+  securityCodePrefix?: string
+}
+
+export interface AttestationServiceSecurityCodeRequest {
```
it would probably make sense to use the same `GetAttestationRequestType` in utils that is used by attestation service. could also update the `AttesationServiceRevealRequest` to the `AttestationRequestType`

aslawson commented on packages/env-tests/package.json:
```diff
@@ -5,7 +5,7 @@
   "license": "MIT",
   "dependencies": {
-    "@celo/contractkit": "0.4.18",
+    "@celo/contractkit": "0.4.19-dev",
```
was this a published version rather than dev intentionally? will be need to change it back after new contractkit version is published?

i1skn replied:
@nambrot can you help us here? Should it be non-dev by intention?

i1skn replied:
Ok, I've reverted the change, so `env-tests` still depends on the released `0.4.18`

nambrot replied:
~~Yeah I was trying to use a published version intentionally to make moving out of the monorepo easier~~

Actually looking at this again, I'd appreciate changing this to the dev-version and making the adjustment on env-tests if you already done so

aslawson commented on packages/mobile/src/identity/securityCode.ts:
```diff
@@ -0,0 +1,103 @@
+  const securityCodePrefix = parseInt(securityCodeWithPrefix[0], 10)
+  const lookupAttestations = attestations.filter(
+    (attestation) => hashAddressToSingleDigit(attestation.issuer) === securityCodePrefix
+  )
```
any thoughts on caching this at the same time that cache ActionableAttestations rather than hashing+filtering all attestations each time someone wants to send a security code to a service?

considering the number of actionable attestations will be low, the time savings it probably not super significant. but seems like the most efficient way to implement.

aslawson commented on packages/mobile/src/identity/securityCode.ts:
```diff
@@ -0,0 +1,103 @@
+    (attestation) => hashAddressToSingleDigit(attestation.issuer) === securityCodePrefix
+  )
+
+  for (const attestation of lookupAttestations) {
```
could consider parallelizing this.

pro - could save time
con - sequential request could reduce the number or requests needed (eg. the match is earlier in the list)

aslawson commented on packages/mobile/src/identity/securityCode.ts:
```diff
@@ -0,0 +1,103 @@
+        TAG + '@getAttestationCodeForSecurityCode',
+        'Getting attestation code for security code error: ',
+        e
+      )
```
wonder if we should differentiate between an expected error "wrong code" and others which we should log as Logger.error?

aslawson reviewed (approved):
looks really good. and love the demo -- so beautiful!

just a couple comments/ideas, but nothing blocking

i1skn committed:
Request validatros in parallel for attestation codes

i1skn commented:
@aslawson @codyborn I've addressed all the feedback. Also, decided to make requests to obtain attestation code in parallel to all matching issuers in case of a collision, but without a retry.

i1skn committed:
Fix build

i1skn committed:
fix celotool build

nambrot commented on packages/mobile/src/identity/securityCode.ts:
```diff
@@ -0,0 +1,119 @@
+  return new BigNumber(address.toLowerCase()).modulo(10).toNumber()
+}
+
+export function getSecurityCodePrefix(issuerAddress: Address) {
```
Should this code not live in ContractKit or Utils? I.e. other wallets will need to do this too right?

i1skn replied:
Well, if other wallet would be implementing the short code verification, than yes. We can move it to `Utils` I think.

aslawson replied:
I think a lot of the current implementation exists in the Attestation Wrapper (though not saying that's the most appropriate place for it...)

aslawson commented on packages/mobile/src/identity/securityCode.ts:
```diff
@@ -1,21 +1,26 @@
+  if (!cache[issuerAddress]) {
+    cache[issuerAddress] = `${hashAddressToSingleDigit(issuerAddress)}`
+  }
+  return cache[issuerAddress]
```
I don't actually think this is providing that much gain, hashAddressToSingleDigit doesn't seem too computationally complex. I was thinking we could have a cache like you originally suggested that was `key:digit --> List<IssuerAddresses>`. That way when you receive a code, you can just check the cache for the list of issuers to send to rather than iterating through each issuer. Though, it would require more maintenance -- eg. removing entries whenever you remove entries from the existing ActiveAttestations cache

i1skn replied:
I would not include it in the current PR. We can handle it later as an improvement to the current solution.

codyborn commented on packages/mobile/src/identity/securityCode.ts:
```diff
@@ -0,0 +1,119 @@
+// for reject and resolve to achieve what we need
+async function raceUntilSuccess<T>(promises: Array<Promise<T>>) {
+  try {
+    const errors = await Promise.all(
```
We can use `Promise.any` and not have to switch the definitions

i1skn replied:
Unfortunately it's not a part of https://github.com/microsoft/TypeScript/blob/master/lib/lib.es2015.promise.d.ts

i1skn committed:
move getSecurityCodePrefix to contractkit

i1skn committed:
Update contractkit docs

i1skn commented:
I've addressed all the feedback. Lmk if there anything else, otherwise I would like to merge it tomorrow before 1.6.0 cut. Thanks!

aslawson reviewed (approved):
shipit 🐑

mergify[bot] committed:
Merge branch 'master' into i1skn/8-digits-verification
